### PR TITLE
Pyupgrade: Format specifiers

### PR DESCRIPTION
**Pull request by colin99d**
A part of #827. Posting this for visibility. Still has some work to do to be done.

### CODE REVIEW

**Comment by colin99d** [comment]
I will try to have this done by Sunday night.

File: src/pyupgrade/plugins/format_specififiers.rs [hunk 1]
```diff
@@ -15,7 +15,7 @@ use crate::cst::matchers::{match_call, match_expression};
 // The regex documentation says to do this because creating regexs is expensive:
 // https://docs.rs/regex/latest/regex/#example-avoid-compiling-the-same-regex-in-a-loop
 lazy_static! {
```
Comment (charliermarsh) [inline src/pyupgrade/plugins/format_specififiers.rs:17]: We use `Lazy` for this, you can grep for some other usages of `Regex` to see an example :)

**Comment by colin99d** [comment]
@charliermarsh I have the following function to detect expressions:
```
pub fn match_expression(expression_text: &str) -> Result<Expression> {
    match libcst_native::parse_expression(expression_text) {
        Ok(expression) => Ok(expression),
        Err(_) => bail!("Failed to extract CST from source"),
    }
}
```
It works fine for an expression like this:
```
print(
    'foo{0}' 'bar{1}'.format(1, 2)
)
```
But the second you make it multi-line like so:
```
print(
    'foo{0}'
    'bar{1}'.format(1, 2)
)
```
The whole thing fails. Is this an error with my coding, or an error with libcst_native?

**Comment by charliermarsh** [comment]
I'm wondering if you need to dedent the code. You might be passing `LibCST` the indented expression, which would cause an error.

**Comment by colin99d** [comment]
It looks like the issue still happens with the de-indented code: 'foo{0}'\n'bar{1}'.format(1, 2).

**Comment by colin99d** [comment]
thread 'main' panicked at 'called `Result::unwrap()` on an `Err` value: ParserError(ParseError { location: ParseLoc { start_pos: LineCol { line: 2, column: 4, offset: 13 }, end_pos: LineCol { line: 2, column: 12, offset: 21 } }, expected: ExpectedSet { expected: {"EOF"} } }, "'foo{0}'\n    'bar{1}'.format(1, 2)")', src/pyupgrade/plugins/format_specififiers.rs:57:57

Looks like this is a parse error from the `parse_expression` function in libcst_native. My thoughts are we either:
1. Make a custom patch in your fork
2. Find a way to modify the string to add a `\` at the end of each line so the parser can understand it, and then remove it once we are done
3. Don't use libcst_native ( I dont like this one because I like libcst now)

Just let me know which you would like for me to pursue.

**Comment by charliermarsh** [comment]
Would you be open to filing an issue on the LibCST repo? In the meantime, we could just skip those cases. (Are they needed for detection, or just autofixing?)

**Comment by colin99d** [comment]
Here is the issue: 
https://github.com/Instagram/LibCST/issues/846
I will just make it a check for now!

File: src/pyupgrade/plugins/format_specififiers.rs [hunk 2]
```diff
@@ -146,16 +147,16 @@ pub fn format_specifiers(checker: &mut Checker, expr: &Expr, func: &Expr) {
...
+                        Diagnostic::new(violations::FormatSpecifiers, Range::from_located(expr));
+                    if checker.patch(diagnostic.kind.code()) {
+                        diagnostic.amend(Fix::replacement(
                             new_call,
                             expr.location,
                             expr.end_location.unwrap(),
                         ));
                     }
-                    checker.add_check(check);
+                    checker.diagnostics.push(diagnostic);
```
Comment (charliermarsh) [inline src/pyupgrade/plugins/format_specififiers.rs:159]: Thank you and sorry that this all churned mid-PR for you.

Comment (colin99d) [reply src/pyupgrade/plugins/format_specififiers.rs]: Im never going to complain about you making this project easier to contribute to.

**Comment by colin99d** [comment]
Hey @charliermarsh I got all the positive test cases working (except for the one edge case we talked about), and all the negative cases working except one. I wanted to get your feedback before I implemented it.
The following edge case is what causes issues:
```
'{' '0}'.format(1)
```
Since the format specifier is split into two different strings, it is NOT supposed to be formatted. However, both the Expr
 struct, and the tokens automatically combine this into one string. What would be the best way of detecting that this edge case is preset. Or do we want to ignore it? Since the refactoring would produce changes that could be seen as desirable:
Screenshot 2023-01-08 at 10 07 52 PM

**Comment by charliermarsh** [comment]
I think it's best just to ignore it (i.e., treat it as if there is no space). It strikes me as exceptionally rare, to the point that it's likely either a mistake _or_ that it's not worth adding more complexity to our own implementation to handle it.

**Comment by charliermarsh** [comment]
We could _probably_ detect it by tokenizing... but it doesn't seem worth it to me.

**Comment by colin99d** [comment]
I just need to find a way to make it cause issues on run and we should be good, will try to have done tonight.

**Comment by colin99d** [comment]
@charliermarsh anything else you want to see from this for merge?

**Comment by charliermarsh** [comment]
@colin99d - Will review this tonight!

File: src/pyupgrade/rules/format_specifiers.rs [hunk 3]
```diff
@@ -0,0 +1,215 @@
...
+            Err(_) => return Err(()),
+            Ok(item) => call.func = Box::new(item),
+        };
+        // Create the string
+        let mut final_state = CodegenState::default();
+        expression.codegen(&mut final_state);
+        // FOR REVIEWER: If the new and old are identical, don't create a fix. Pyupgrade
+        // doesnt even want me to report this, so we could just have an enum for errors,
+        // and if a special one is returned here then we dont even report a fix
+        if module_text == final_state.to_string() {
```
Comment (charliermarsh) [inline src/pyupgrade/rules/format_specifiers.rs:97]: When does this happen?

File: src/pyupgrade/rules/format_literals.rs [hunk 4]
```diff
@@ -0,0 +1,118 @@
...
+    expression.codegen(&mut state);
+    if module_text == state.to_string() {
+        // Ex) `'{' '0}'.format(1)`
+        bail!("Failed to generate call expression for: {module_text}")
+    }
+    Ok(state.to_string())
+}
+
+/// UP030
+pub(crate) fn format_literals(checker: &mut Checker, summary: &FormatSummary, expr: &Expr) {
```
Comment (charliermarsh) [inline src/pyupgrade/rules/format_literals.rs:85]: @colin99d - It turns out that we already had a utility for extracting the format positions, which uses the RustPython string parser underneath and so is very robust. Sorry that I didn't flag this earlier -- I didn't realize it could even be reused here, but it should make things more efficient too since we can do one string parse and share that "summary" amongst a bunch of checks.

File: src/pyupgrade/rules/format_literals.rs [hunk 5]
```diff
@@ -0,0 +1,118 @@
...
+            &checker
+                .locator
+                .slice_source_code_range(&Range::from_located(expr)),
+            &summary.indexes,
+        ) {
+            diagnostic.amend(Fix::replacement(
+                contents,
+                expr.location,
+                expr.end_location.unwrap(),
+            ));
```
Comment (charliermarsh) [inline src/pyupgrade/rules/format_literals.rs:114]: I feel like one strategy that could work here would be...

- Take the entire text.
- Replace the string part via the regex above.
- Replace the arguments via LibCST.

Kinda tough, hacky, etc... but would solve the missing cases.

Comment (charliermarsh) [reply src/pyupgrade/rules/format_literals.rs:114]: Actually, I guess that wouldn't solve the `'{' '0}'.format(1)` case.

**Comment by colin99d** [comment]
Thanks for the review!
